### PR TITLE
Pure-python masked UDFs

### PR DESCRIPTION
**Pull request by brandon-b-miller**
Replaces C++ implementation of masked UDF pipeline with a pure python version which compiles and launches the entire kernel using numba. This solves a bunch of problems:

- CUDA 11.0 support is now available since the impl no longer needs `cuda::std::tuple` to work with NVRTC 11.0. 
- Support for special functions which compile to multiple function definitions, such as `pow`, `sin`, and `cos` is now provided since all the PTX is compiled and linked inside numba (Fixes https://github.com/rapidsai/cudf/issues/8470)
- Allows us to support this corner case which would require a separate c++ kernel in previous implementation
```python
def f(x):
    return 42
```

- Makes developing/adding features to the impl much easier

### CODE REVIEW

**Comment by brandon-b-miller** [comment]
cc @gmarkall

**Comment by brandon-b-miller** [comment]
This should be ready for review. 

cpp folks - what should be the fate of this c++ code? There's an old issue around an `n-ary` transform in libcudf that could maybe be largely satisfied by what is here, modulo some tests that don't exist. That said I think the only use case that this was supporting was the original pipeline.

**Comment by jrhemstad** [comment]
> cpp folks - what should be the fate of this c++ code? There's an old issue around an `n-ary` transform in libcudf that could maybe be largely satisfied by what is here, modulo some tests that don't exist. That said I think the only use case that this was supporting was the original pipeline.

Python has been the only user of any of this stuff. It's the only front-end that has something like Numba to generate PTX code. We'd probably want to spend a little time talking about it, but I'd be supportive of just removing all of it all together if it's no longer going to be used. No sense in maintaining something that isn't used.

File: python/cudf/cudf/core/udf/pipeline.py
Comment (shwina) [inline python/cudf/cudf/core/udf/pipeline.py:21]: We should change the docstring of this function since the PTX is no longer handled by libcudf.

File: python/cudf/cudf/core/udf/pipeline.py
Comment (shwina) [inline python/cudf/cudf/core/udf/pipeline.py:76]: docstring

File: python/cudf/cudf/core/udf/pipeline.py
Comment (shwina) [inline python/cudf/cudf/core/udf/pipeline.py:180]:
```suggestion
@annotate("UDF PIPELINE", color="black", domain="cudf_python")
```

`black` is unfortunately not a color we support (unless matplotlib is installed). I believe it would also be difficult (impossible) to read in the profiles.

https://github.com/NVIDIA/NVTX/blob/dev/python/nvtx/colors.py#L9

File: python/cudf/cudf/core/udf/pipeline.py
Comment (shwina) [inline python/cudf/cudf/core/udf/pipeline.py:134]: `.columns` is expensive -- use` ._data` instead (bonus: works for `Series` as well)

File: python/cudf/cudf/core/udf/pipeline.py
Comment (shwina) [inline python/cudf/cudf/core/udf/pipeline.py:152]:
```suggestion
        initializer = template.format(idx=idx)
```

File: python/cudf/cudf/core/frame.py
Comment (shwina) [inline python/cudf/cudf/core/frame.py:1344]: I think you need:

```
import cudf.core.udf.pipeline
```

just before this (I did).

File: python/cudf/cudf/core/udf/pipeline.py
Comment (shwina) [inline python/cudf/cudf/core/udf/pipeline.py:17]: Nit: constants are typically `ALL_CAPS`

File: python/cudf/cudf/core/udf/pipeline.py
Comment (shwina) [inline python/cudf/cudf/core/udf/pipeline.py:195]: What is `lcl`? Assuming the local namespace?

**Review by gmarkall — CHANGES_REQUESTED** [review]
Congratulations on getting this all to work together with Numba and free up the UDF machinery from the C++ side of things! Having looked at this, there are a few comments on the diff and some larger questions:

First, do we know if this has any impact on performance? Theoretically we should be able to generate very similar PTX / SASS from Numba as we can for CUDA C / C++, but if the typing does something surprising / unhelpful, sometimes this can impact performance, especially by reducing occupancy by increasing register usage because of wider types.

Secondly, I had a thought about how we could eliminate the whole step of compiling the "inner" UDF `f` to get the return type to know how to handle it. If we can't do this, there are some comments on the diff about how I think we could make the process a bit more efficient in terms of caching, but perhaps there's another way around this. Instead of figuring out whether the return value (`ret` in the kernel template) is masked or unmasked, we could instead implement a function (e.g. call it `unpack_return`) that has two different typings and lowerings, one for a scalar return value, and one for a masked return value.

For the scalar return value, its implementation would be equivalent to:

```python
@cuda.jit(device=True)
def unpack_return(ret):
    return ret, True
```

For a masked return value, its implementation would be equivalent to:

```python
@cuda.jit(device=True)
def unpack_return(ret):
    return ret.value, ret.valid
```

then in the kernel template, instead of 

```python
        ret_data_arr[i] = {ret_value}
        ret_mask_arr[i] = {ret_valid}
```

we'd have:

```python
        ret_value, ret_valid = unpack_return(ret)
        ret_data_arr[i] = ret_value
        ret_mask_arr[i] = ret_valid
```

and in `udf_pipeline`, we'd be able to remove the call to `compile_udf` and everything relating to `scalar_return_type`.

This would save a needless compilation, and reduce the complexity of the UDF pipeline, at the cost of implementing the typings and lowerings for the `unpack_return` function (which I think is better in terms of lowering overall complexity and redundancy). What do you think of this idea? I'm happy to discuss / expand on it in a more interactive form if that would be helpful.

Finally - the UDF pipeline is doing something quite complex, and even though I wrote some of the code for it, I did find it quite hard to check / follow. I think it would be really helpful for future readers if there were some comments (possibly close to the kernel template) that describes the interface / contract between the generated kernel, and the code that calls it - including a description of the arguments, their order, the purpose of each argument and expected type, roughly what the kernel does (calls the inner UDF for each element etc.), and what `udf_pipeline` is doing to prepare the arguments and package the output back up for returning.

File: python/cudf/cudf/core/frame.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/frame.py]: Should be fixed

File: python/cudf/cudf/core/udf/pipeline.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/pipeline.py]: Fixed

File: python/cudf/cudf/core/udf/pipeline.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/pipeline.py:134]: Fixed

File: python/cudf/cudf/core/udf/pipeline.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/pipeline.py]: did this :)

File: python/cudf/cudf/core/udf/pipeline.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/pipeline.py]: this should be resolved

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_udf_masked_ops.py]: removed

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_udf_masked_ops.py:99]: Pandas still returns something is the problem. These are all cases where pandas ultimately defers to python scalar logic whereas we defer to libcudf. There may come a day where these actually do pass - but that basically requires dataframe APIs to be standardized across the python ecosystem. @shwina has more info on this specifically but as of now I don't think there's a consensus on answers to these questions.

File: python/cudf/cudf/utils/cudautils.py
Comment (brandon-b-miller) [reply python/cudf/cudf/utils/cudautils.py]: Changed this slightly - this isnt really needed but somewhat reduces code duplication.

**Comment by brandon-b-miller** [comment]
@gmarkall I implemented `unpack_return` a little differently - I found it was easier to write typing and lowering for a function `pack_return` that more or less does the inverse, but is still to the same ends. Just moving our PM discussion to github - it seems that we still cant avoid precompiling the inner function because we need it to figure out what kind of output column to allocate for the kernel. That said, each function uses its own cache, so I believe at least that an already used kernel shouldn't need to recompile anything.

File: python/cudf/cudf/core/udf/pipeline.py
Comment (gmarkall) [inline python/cudf/cudf/core/udf/pipeline.py:182]: Referring to the previous comment https://github.com/rapidsai/cudf/pull/9174/#discussion_r707244778 which I can't reply to now it's not on this diff: I think the suggestion in that comment is only partially applied, because the code here still calls `compile_masked_udf` before checking the cache using a key made up of the argument dtypes and the code and closure variables of the function.

However, I recall you mentioned yesterday in an OOB conversation that there's still some work to be done here, because the cache key also depends on whether the input is masked, and not just what its dtype is, because kernels are different depending on whether the inputs are masked or not (whether they create a dummy "True" mask or not) - so I think there's further changes to go in in this are at the moment anyway?

**Review by gmarkall — CHANGES_REQUESTED** [review]
Many thanks for the updates! I see now that we still can't get rid of the compilation entirely yet, because we need to type the inner function to know what to pass in as the argument for the return type to the global kernel.

The `pack_return` implementation looks like a good idea. I've made some additional comments on the diff.

One other thing from the previous review that I think is still outstanding is a bit more documentation (possibly in comments) outlining the interface between the generated kernels and the cuDF runtime, though I understand some of this is still subject to change so you might have been holding off until it was a bit more finalised first.

File: python/cudf/cudf/core/udf/pipeline.py
Comment (gmarkall) [reply python/cudf/cudf/core/udf/pipeline.py]: Can you expand on what you mean / provide an example that triggers this? I'm wondering if there's a Numba bug here...

File: python/cudf/cudf/core/udf/typing.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/typing.py]: Done

File: python/cudf/cudf/core/udf/pipeline.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/pipeline.py]: Yup - updated

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_udf_masked_ops.py:99]: I see what you mean. From that perspective I suppose there are two options:

- Detect the troublesome parameter combinations during testing and skip them
- Write a parameter generating function that never generates those test cases in the first place

I will probably go with the first one for simplicity's sake unless anyone feels strongly that we go another route.

File: python/cudf/cudf/utils/cudautils.py
Comment (brandon-b-miller) [reply python/cudf/cudf/utils/cudautils.py]: Updated this

**Review by robertmaynard — APPROVED** [review]
CMake changes LGTM

**Comment by brandon-b-miller** [comment]
Upon consulting with @shwina we decided it was better to leave the c++ alone in this PR and open a separate one removing the c++ side of things. In that case it could be reverted separately without having to revert the new pipeline as well.

**Comment by brandon-b-miller** [comment]
This should be ready to go.

File: python/cudf/cudf/core/udf/lowering.py
Comment (galipremsagar) [inline python/cudf/cudf/core/udf/lowering.py:14]: Can we make these imports absolute to be consistent with the rest of the code-base?

File: python/cudf/cudf/core/udf/typing.py
Comment (galipremsagar) [inline python/cudf/cudf/core/udf/typing.py:21]: Absolute imports here too.

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/cudf/core/udf/pipeline.py
Comment (gmarkall) [reply python/cudf/cudf/core/udf/pipeline.py:135]: Does that mean that this code is now dead? Or is an update here needed too?

File: python/cudf/cudf/core/udf/pipeline.py
Comment (gmarkall) [inline python/cudf/cudf/core/udf/pipeline.py:16]: I believe this is superfluous - since you also implement typing and lowering for `pack_return`, it doesn't need to be jitted (and if this jitted version ever got used, it would not do anything because it jits a function with an empty body).

**Review by gmarkall — CHANGES_REQUESTED** [review]
I think this is now looking pretty good - there are some small suggestions on the diff but they're fairly minor. I wanted to double-check the suggestions made sense so I had to implement them - if you like them, you could just pull in the last three commits from https://github.com/gmarkall/cudf/tree/masked-udfs-20210927 which implement the suggestions on the diff here.

**Review by gmarkall — APPROVED** [review]
LGTM!

**Comment by brandon-b-miller** [comment]
rerun tests

File: python/cudf/cudf/utils/cudautils.py
Comment (shwina) [inline python/cudf/cudf/utils/cudautils.py:257]: Let's also move the 2 comments above into the `make_cache_key` function instead, as a "docstring".

**Review by shwina — APPROVED** [review]

**Comment by shwina** [comment]
@gpucibot merge

**Comment by shwina** [comment]
This is awesome work, @brandon-b-miller!

**Pull request merged**
